[#13] 회원 정보 수정 기능

haileyKimDev opened the pull request:
## 개요
회원 정보 수정 기능을 추가하였습니다.

## 작업사항
- 회원정보 수정화면 추가
- 유저정보 수정 로직 추가
- 유저 정보 변경 성공 테스트케이스 추가 
- related : #13

## 변경로직
- [user_edit.html](https://github.com/f-lab-edu/s-mart/compare/develop...feature/13-edit-userinfo?expand=1#diff-e93e27720f2a7470ea8f0cad0f6fdb4b871220b44ad51daa4e8879e612514f6f) 회원정보 수정화면 페이지 및 스크립트 추가
- [UserMapper.xml](https://github.com/f-lab-edu/s-mart/compare/develop...feature/13-edit-userinfo?expand=1#diff-02cb8fe615400adf318536db1b0ce8306082a295569c9db02fbfa7dc9acffcef) 회원정보 수정 쿼리 추가
- [UserServiceImpl.java](https://github.com/f-lab-edu/s-mart/compare/develop...feature/13-edit-userinfo?expand=1#diff-26074b33aef98e20eefed462375288f623b12058257a1a83b8f897483b791057) 회원정보 수정 메소드 구현 
- [UserService.java](https://github.com/f-lab-edu/s-mart/compare/develop...feature/13-edit-userinfo?expand=1#diff-435187282ceb4404c868206f8ca8d6b6fbd023d3ac049bf3aff40eecba2e8cce) 회원정보 수정 메소드 및 닉네임 중복 체크 메소드 추가
- [UserViewController.jav](https://github.com/f-lab-edu/s-mart/compare/develop...feature/13-edit-userinfo?expand=1#diff-0474330c54e34ff6dff4cba92cc1305070b85133ada694a78a39a8a974df655a) 뷰컨트롤러에 회원정보수정 추가
- [UserController.java](https://github.com/f-lab-edu/s-mart/compare/develop...feature/13-edit-userinfo?expand=1#diff-163e99b2a9617d3dd5fbd827b33705ed04a4e3902e199f46bc3e3cefa3ea7a57) 닉네임중복 api, 회원정보변경 api 추가 
- [UserServiceImplTest.java](https://github.com/f-lab-edu/s-mart/compare/develop...feature/13-edit-userinfo?expand=1#diff-76bfc33698bac02ef5dd919dbf18f9f38ed6fcc3549c2940a6559fea9b56c8b4) 유저정보변경 테스트 코드 추가

haileyKimDev committed:
feat: 사용자 정보 수정 기능 구현
- 회원정보 수정화면 추가
- 유저정보 수정 로직 추가
- 유저 정보 변경 성공 테스트케이스 추가
related : #13

f-lab-fi commented on src/main/java/com/smart/user/controller/UserController.java:
```diff
@@ -53,4 +55,17 @@ public ResponseEntity<?> join(@RequestParam String email, @RequestParam String a
+    return ResponseEntity.ok(isUpdated);
+  }
+
+  @GetMapping("/checkNickname")
+  public ResponseEntity<Boolean> checkNickname(@RequestParam String nickname) {
+    //닉네임 중복체크
+    boolean isDuplicated = userService.isDuplicateNickName(nickname);
+    return ResponseEntity.ok(isDuplicated);
+  }
```
두가지 일관성이 없는 부분이 있습니다. :)

1. /checkNickname 에서는 Nick**n**ame으로 적고, isDuplicateNickName 에서는 Nick**N**ame으로 적음.
2. /join-auth 에서는 하이픈으로 구분해줬는데, /checkNickname는 대소문자로 구분.

f-lab-fi replied:
그리고 일부 API를 제외하면 API를 Restful 하게 만들 수 있을 것 같은데, 한번 찾아보시고 수정해보시면 좋을 것 같습니다. (_ _)

haileyKimDev replied:
@f-lab-fi 

1. 카멜케이스 형식을 통일하였습니다.
2. 하이픈 구분으로 url 변경하였습니다.
3. 각 리턴을 서비스레이어의 반환값에 따라 상태코드 및 메세지를 담아 리턴하도록 변경하였습니다 !

haileyKimDev committed:
refactor: 카멜케이스 규칙 동일하게 적용
related: #13

haileyKimDev committed:
refactor: 회원가입 컨트롤러 변경
http 상태를  ResponseEntity리턴시에 담도록 변경
회원가입 완료후  유저정보를 바디에 담아 리턴
related: #13

haileyKimDev committed:
refactor: 이메일 인증 컨트롤러 정정
인승실패시 상태코드 400 리턴으로 변경
related: #13

haileyKimDev committed:
refactor: 리턴형태 변환
서비스 레이어의 응답값에 따라 상태코드 및 메시지변경
related: #13

f-lab-fi commented on src/main/java/com/smart/user/controller/UserController.java:
```diff
@@ -41,31 +38,38 @@ public ResponseEntity<UserInfo> getUser(@PathVariable("email") String email) {
+      return ResponseEntity.badRequest()
+          .body("이메일 인증에 실패했습니다.");
+    }
   }
```
[userService.verifyAuthCode의 반환값이 void에서 boolean으로 변경](https://github.com/f-lab-edu/s-mart/pull/19/files/d7db6f058319ba0a463eaab7e3d18b447ec0d2e1..8519a22d440b3bc28503cb85351c98d7a18b0d64#diff-435187282ceb4404c868206f8ca8d6b6fbd023d3ac049bf3aff40eecba2e8cceL11-R11)되었는데요, 이렇게 반환값을 받아서 처리하기 보다는 userService.verifyAuthCode에서 예외를 던지는 방법으로 처리했으면 어땠을까요?

정상적으로 인증되는 경우 예외를 던지지 않고, 인증에 실패한 경우에는 예외를 던지는겁니다.
예외에 대한 핸들링은 전역 예외처리 핸들러(@RestControllerAdvice)를 통해서 해주고요.

"이메일 인증이 완료되었습니다."와 "이메일 인증에 실패했습니다." 같은 body가 없더라도 상태코드로 충분히 인증 실패 여부를 알려줄 수 있을 것 같습니다. :)
만약 body가 필요하다고 해도 전역 예외처리 핸들러에서 처리해주면 될거구요.

haileyKimDev replied:
@f-lab-fi 넵 멘토님! 
서비스에서 예외를 던지고 있으면서도 반환타입을 불리언 리턴하고있었는데,
불필요하게 리턴 처리를 했던것 같습니다. 
말씀해주신대로 서비스에서 인증실패시에 예외를 던지고 @RestControllerAdvice를 붙인 ExceptionAdvice 클래스를 생성하였습니다!

f-lab-fi commented on src/main/java/com/smart/user/controller/UserController.java:
```diff
@@ -41,31 +38,38 @@ public ResponseEntity<UserInfo> getUser(@PathVariable("email") String email) {
-    return ResponseEntity.ok(isUpdated);
+    return ResponseEntity.ok()
+        .body(isUpdated);
   }
```
요것도 굳이 응답에 isUpdated에 대한 true or false 값이 존재할 필요는 없어보입니다. :)

haileyKimDev replied:
닉네임변경 메소드도 불필요한 리턴값을 없애고
서비스단에서 닉네임 중복값을 확인하고 예외를 던지도록 변경하였습니다!

f-lab-fi commented on src/main/java/com/smart/user/controller/UserController.java:
```diff
@@ -41,31 +38,38 @@ public ResponseEntity<UserInfo> getUser(@PathVariable("email") String email) {
+      return ResponseEntity.status(HttpStatus.CONFLICT).build();
+    } else {
+      return ResponseEntity.ok().build();
+    }
```
https://github.com/f-lab-edu/s-mart/pull/19/files/d7db6f058319ba0a463eaab7e3d18b447ec0d2e1..8519a22d440b3bc28503cb85351c98d7a18b0d64#r1176760364

위 리뷰와 마찬가지로, 예외로 처리해보시기 바랍니다!

haileyKimDev replied:
닉네임중복여부 api도
서비스에서 중복여부를 체크하고 중복시 예외를 던지도록 변경하였습니다 !

f-lab-fi commented on src/main/java/com/smart/user/service/implement/UserServiceImpl.java:
```diff
@@ -60,14 +62,16 @@ private void saveAuthCode(String email, String authCode) {
-      userDao.updateUserStatus(email, Status.NORMAL);
-    } else {
+  public boolean verifyAuthCode(String email, String authCode) {
+    String storedAuthCode = (String) session.getAttribute(email);
```
그러고보니 요거 `HttpSession`으로 구현하셨었군요.. 가급적 Service 계층은 표현 계층의 HTTP를 모르도록 만들어보시는게 좋습니다. :)
인증 코드를 관리하기 위한 Repository를 추가해보시기 바랍니다.
우선 개발을 위해서는 Map 같은걸로 구현하시면 좋을 것 같습니다.

haileyKimDev replied:
넵! 
AuthCodeRepository 인터페이스 생성 후 AuthCodeRepositoryImpl 구현체에서 
인증코드 관련 메소드를 옮기고 Map으로 처리하였습니다!

haileyKimDev committed:
refactor: 서비스에서 예외던지도록 변경
예외처리 핸들러 사용예정으로 컨트롤러에서의 분기처리를 없앰
related: #13

haileyKimDev committed:
refactor: UserDao를 UserRepository 로 변경
- UserDao 인터페이스를 UserRepository로 추상화
- AuthCodeRepository 추가
- ExceptionAdvice 추가
- UserController 변경

haileyKimDev committed:
refactor: UserDao를 UserRepository 로 변경
- UserDao 인터페이스를 UserRepository로 추상화
- AuthCodeRepository 추가
- ExceptionAdvice 추가
- UserController 변경

haileyKimDev committed:
refactor: AuthCodeRepository 추가
- AuthCodeRepository추가로 기존 session 로직제거

f-lab-fi commented on src/main/java/com/smart/user/repository/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,47 @@
+
+  private static Long sequence = 1L;
+
+  @Override
+  public Optional<User> findByEmail(String email) {
+    return users.stream()
+        .filter(user -> user.getEmail().equals(email)).findFirst();
+  }
+
+  @Override
+  public void deleteByEmail(String email) {
+    this.findByEmail(email).ifPresent(users::remove);
+  }
```
```
.filter(user -> user.getEmail().equals(email)).findFirst();

this.findByEmail(email).ifPresent(users::remove);
```

요런것들 메서드 단위로 개행 해주세요~
현재 상태는 가독성이 떨어집니다.

haileyKimDev replied:
넵! 조금 나눠서 개행하도록 변경하였습니다!

f-lab-fi commented on src/main/java/com/smart/user/domain/User.java:
```diff
@@ -22,4 +22,24 @@ public class User {
+    this.userId = userId;
+  }
+
+  public void setUserStatus(Status userStatus) {
+    this.userStatus = userStatus;
+  }
+
+  public void setName(String name){
+    this.name = name;
+  }
+
+  public void setPassword(String password){
+    this.password = password;
+  }
+
+  public void setNickname(String nickname){
+    this.nickname = nickname;
+  }
```
setXXX, getXXX 같은 메서드는 가급적 지양하는게 좋습니다.
물론 필연적으로 사용해야하는 경우도 있지만, 의미 있는 메서드 이름을 붙여주세요.

https://github.com/f-lab-edu/s-mart/pull/19/files/8519a22d440b3bc28503cb85351c98d7a18b0d64..5ffb1fc759aba8a4fad59a94ad28a2e910f19161#diff-435187282ceb4404c868206f8ca8d6b6fbd023d3ac049bf3aff40eecba2e8cceR61

위 코드를 보면, `user.setUserStatus(Status.NORMAL);` 라고 되어있는데, 여기서 이 코드를 읽는 사람이 UserState를 NORMAL로 바꾼다는게 무엇을 의미하는지 파악하기 쉬울까요~?

haileyKimDev replied:
네! 어떤 행위를 하는지 메소드명으로 표시하도록 하겠습니다.
updateUserStatus로 변경하였습니다 !

f-lab-fi commented on src/test/java/com/smart/user/repository/UserRepositoryTest.java:
```diff
@@ -0,0 +1,62 @@
+    userRepository.deleteByEmail(user.getEmail());
+  }
+
+  @Test
+  public void 유저회원가입() {
+    userRepository.save(user);
+    Optional<User> foundUser = userRepository.findByEmail(user.getEmail());
+    Assertions.assertTrue(foundUser.isPresent());
+  }
+
+  @Test
+  public void 중복이메일체크() {
+    userRepository.save(user);
+
+    boolean checkUserMail = userRepository.existsByEmail("test@email");
+    Assertions.assertTrue(checkUserMail);
+
+    checkUserMail = userRepository.existsByEmail("not-in@email");
+    Assertions.assertFalse(checkUserMail);
+  }
+
+  @Test
+  public void 중복닉네임체크() {
+    userRepository.save(user);
+
+    boolean checkUserNickname = userRepository.existsByNickname("nickname");
+    Assertions.assertTrue(checkUserNickname);
+
+    checkUserNickname = userRepository.existsByNickname("not-in");
+    Assertions.assertFalse(checkUserNickname);
+  }
```
테스트 이름을 통해 (혹은 @DisplayName) 무엇을 테스트 하려는거고, 어떤 상황에 성공하고 어떤 상황에 실패하는건지 명시하는게 좋을 것 같습니다.

추가적으로 중복이메일체크, 중복닉네임체크 두 테스트는 한번에 2가지가 테스트 되고 있어보여요.
중복이 되는 것과 안되는 것은 별개로 봐야 적절할 것 같습니다.
아마 코드를 잘못 수정해서 테스트가 깨지는 상황이라면, 모든 값이 중복 된다고 나오거나, 모든 값이 중복되지 않는다고 나올텐데, 이 때 테스트가 구분되어 있어야 더 빠르게 파악할 수 있을겁니다.

haileyKimDev replied:
닉네임이 중복되는 경우와 중복이 아닌경우를 모두 테스트에 넣었었네요 !
중복인경우와 중복이아닌경우로 나누어서 테스트코드를 다시 구현했습니다 !

haileyKimDev committed:
refactor : 메소드명 변경 및 개행
related : #13

haileyKimDev committed:
refactor : develop 브랜치와 병합
related: #13

haileyKimDev committed:
refactor : security 계층 분리로 import 정정
related : #13

f-lab-fi reviewed (approved):
수고하셨습니다. :)
